cilium/CFP-30984-dns-proxy-ha-v2.md

```diff
@@ -0,0 +1,127 @@
+
+
+### High Level Information Flow
+
+* Agent starts up with gRPC streaming service (only after resources are synced from k8s and ipcache bpf map is populated)
+* SDP starts up.
+* Connects to gRPC service, retrying periodically until success.
+* Agent sends current snapshot for L7 DNS Policy enforcement via UpdatesDNSRules to SDP.
+* On policy recomputation,  agent invokes UpdatesDNSRules.
+* On DNS request from the client, DNS request redirects to DNS proxy port.
+* Kernel round robin load balances between SDP and built in proxy.
+* Assuming SDP gets the request, SDP enforces L7 DNS policy.
+  * Lookup identity based on IP address via bpf map.
+  * Check against policy snapshot if this identity is allowed to resolve the current DNS name and is allowed to talk to DNS server target identity (also needs lookup).
+* Make upstream DNS request from SDP.
+* On response, SDP invokes UpdatesMappings() to notify agent of new mappings.
+* Release DNS response after success from  UpdatesMappings() / timeout.
+
+### Handling SDP <> Agent re-connections
+
+* When the agent is unavailable, SDP will periodically attempt to re-connect to the streaming service. Any FQDN<>IP mappings resolved when the agent is down will be cached in SDP and `UpdatesMappings` will be retried after establishing the connection.
+  * A new bpf map for ipcache is populated on agent startup, so SDP needs to re-open the ipcache bpf map when the connection is re-established. See https://github.com/cilium/cilium/pull/32864 for similar handling in envoy.
+  * On a new connection from SDP, the agent will invoke `UpdatesDNSRules` to notify SDP of all L7 DNS policy rules.
+
+* SDP will not listen on the DNS proxy port until a connection is established with cilium agent and initial L7 DNS policy rules are received. Meanwhile, built-in DNS proxy will continue to serve requests. SDP relies on cilium agent for initial bootstrap. In future, we could make SDP retrieve initial policy information from other sources, but this is not in scope for this CFP.
+
+### Handling Upgrades
+
+Other than the streaming API from the agent, this CFP introduces a dependency on the ipcache bpf map which isn't a stable API exposed to components beyond the agent. Sufficent tests will be added to catch such datapath changes impacting SDP. In order to support a safe upgrade path, SDP would need to support reading from the current and future formats of the map (including possibly reading from an entire new map).
```

Comment:
I think this is a reasonable state of intent / goals, but there is a lot of detail missing here. 

- What are "sufficient tests"?
- What upgrade path do you intend to support?
- What assumptions are you making about the formats of the map? How do you intend to detect if the format is different?

Reply:
@joestringer  reading the BPF map is a common issue for both Envoy and SDP. I hope this CFP doesn't introduce any new dependencies. I'm not pointing to Envoy as an excuse; instead, we should work together to solve this issue for both, which can be a separate discussion. One option we discussed was to abstract the BPF map internals with a Wrapper API that Envoy and SDP can use. This way, they wouldn't need to worry about underlying improvements or changes, and the Wrapper API would remain unaffected

Reply:
> > What assumptions are you making about the formats of the map? How do you intend to detect if the format is different?
> 
> Actually we already have a layer of indirection from the userspace cache and from SDP we're reusing the same cache structure. So any changes to the map, will come with accompanying userspace cache format changes, which SDP would also pick. So this would cover our needs from a format change perspective ?

Sorry, I don't quite follow. Are both SDP and cilium-agent making BPF syscalls to interpret BPF map formats? If yes, what expectations do you have if we want to change the ipcache format in future?

I think we've discussed earlier some potential expectations and solutions, but right now I'm finding it difficult to understand exactly what we're agreeing to. If we agree that this is a future problem, I'm fine with that, just know that means we do not have consensus about exactly how upgrades will behave, and that may lead to temporary packet loss when components are replaced. As long as we're on the same page and document this, I am fine. What I want to avoid is a situation where y'all have high standards for expectations of behavior during upgrade, but we never wrote down exactly what those expectations were and didn't get agreement from other developers it could affect.

Reply:
> SDP would need to support reading from the current and future formats of the map

So I think that this does capture the essence of the proposed solution for upgrades.

@hemanthmalla , @tamilmani1989 and I discussed out-of-band how this could work:

> today there's only one ipcache format and no need for a wrapper. But let's say we want to change the ipcache format. We introduce this wrapper which understands two entirely different map formats and figures out the right format to read, based on some heuristic. Maybe key/value size or something. Then we carry that format change logic as a library for a release or two. If we also upgrade SDP first, it will migrate onto the wrapper and continue to work with the old ipcache ... if you then upgrade cilium, cilium will migrate to the newer ipcache format and the SDP will auto-detect / start reading the new format